File: customer-search-poc/frontend/src/services/customerService.ts
import { Customer } from '../types/Customer';

const API_BASE_URL = '/api';

export interface SearchResponse {
  query: string;
  results: Customer[];
}

export class CustomerService {
  /**
   * Search for customers by query term (or list all if query is empty)
   */
  static async searchCustomers(query: string, page: number = 1): Promise<Customer[]> {
    // Allow empty query for "view all" functionality
    const response = await fetch(
      `${API_BASE_URL}/search?q=${encodeURIComponent(query)}&page=${page}`
    );

    let results: Customer[] = [];

    if (response.ok) {
      const data: SearchResponse = await response.json();
      results = data.results || [];
    } else {
      const error = await response.json();
      throw new Error(error.error || 'Search failed');
    }

    return results;
  }
}

File: customer-search-poc/flutter/lib/services/customer_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/customer.dart';

class CustomerService {
  // API base URL - matches React frontend configuration
  // For web: uses relative path '/api' (same as React)
  // For mobile: update to your server's IP address (e.g., 'http://192.168.1.100:3000/api')
  static const String apiBaseUrl = '/api';

  static Future<List<Customer>> searchCustomers(String query, int page) async {
    try {
      final uri = Uri.parse('$apiBaseUrl/search')
          .replace(queryParameters: {
        'q': query,
        'page': page.toString(),
      });

      final response = await http.get(uri);

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        final results = data['results'] as List<dynamic>;
        return results.map((json) {
          try {
            return Customer.fromJson(json as Map<String, dynamic>);
          } catch (e) {
            print('Error parsing customer: $json, Error: $e');
            rethrow;
          }
        }).toList();
      } else {
        final error = json.decode(response.body);
        throw Exception(error['error'] ?? 'Search failed');
      }
    } catch (e) {
      throw Exception('Failed to load customers: $e');
    }
  }
}
